src/tools/r2rdump/EHInfo.cs

```diff
@@ -0,0 +1,234 @@
+        /// The length of the 
+        /// </summary>
+        /// <param name="image">R2R PE image</param>
+        /// <param name="ehInfoRva">RVA of the EH info</param>
+        /// <param name="methodRva">Starting RVA of the runtime function</param>
+        /// <param name="offset">File offset of the EH info</param>
+        /// <param name="clauseCount">Number of EH info clauses</param>
+        public EHInfo(byte[] image, int ehInfoRva, int methodRva, int offset, int clauseCount)
+        {
+            EHInfoRVA = ehInfoRva;
+            MethodRVA = methodRva;
+            EHClauses = new EHClause[clauseCount];
+            for (int clauseIndex = 0; clauseIndex < clauseCount; clauseIndex++)
+            {
+                EHClauses[clauseIndex] = new EHClause(image, offset + clauseIndex * EHClause.Length);
+            }
+        }
+
+        /// <summary>
+        /// Emit the textual representation of the EH info into a given writer.
+        /// </summary>
+        public void WriteTo(StringBuilder stringBuilder)
+        {
+            foreach (EHClause ehClause in EHClauses)
+            {
+                ehClause.WriteTo(stringBuilder, MethodRVA);
+                stringBuilder.AppendLine();
+            }
+        }
+    }
+
+    /// <summary>
+    /// Location of the EH clauses for a single runtime function
+    /// </summary>
+    public class EHInfoLocation
+    {
+        /// <summary>
+        /// RVA of the EH clause list in the R2R PE image
+        /// </summary>
+        public readonly int EHInfoRVA;
+
+        /// <summary>
+        /// Number of EH clauses
+        /// </summary>
+        public readonly int ClauseCount;
+
+        public EHInfoLocation(int ehInfoRva, int clauseCount)
+        {
+            EHInfoRVA = ehInfoRva;
+            ClauseCount = clauseCount;
+        }
+    }
+
+    /// <summary>
+    /// Lookup table maps IP RVAs to EH info.
+    /// </summary>
+    public class EHLookupTable
+    {
+        /// <summary>
+        /// Map from runtime function RVA's to EH info location in the R2R PE file.
+        /// </summary>
+        public readonly Dictionary<int, EHInfoLocation> RuntimeFunctionToEHInfoMap;
+
+        /// <summary>
+        /// Reads the EH info lookup table from the PE image.
+        /// </summary>
+        /// <param name="image">R2R PE image</param>
+        /// <param name="offset">Starting offset of the EH info in the PE image</param>
+        /// <param name="length">Byte length of the EH info</param>
+        public EHLookupTable(byte[] image, int offset, int length)
+        {
+            RuntimeFunctionToEHInfoMap = new Dictionary<int, EHInfoLocation>();
+
+            int rva1 = BitConverter.ToInt32(image, offset);
+            int eh1 = BitConverter.ToInt32(image, offset + sizeof(uint));
+
+            while ((length -= 2 * sizeof(uint)) >= 8)
+            {
+                offset += 2 * sizeof(uint);
+                int rva2 = BitConverter.ToInt32(image, offset);
```

Comment:
I don't see rva2 being used anywhere. Is that intentional?

Reply:
Of course not, thanks for spotting! This is supposed to be a state machine iterating over successive pairs of records so that, at the end of the loop statement, we need to assign rva2 -> rva1, eh2 -> eh1. Will fix!